scalameta/tokenizers/src/main/scala/scala/meta/internal/tokenizers/XmlParser.scala

```diff
@@ -0,0 +1,161 @@
+  ))
+  val Ideographic = P( CharIn(
+    '\u4E00' to '\u9FA5',  "\u3007", '\u3021' to '\u3029'
+  ))
+  // format: on
+  val Eq = P(WL.? ~ "=" ~ WL.?)
+  val ScalaExpr = P("{" ~ WS ~ Block ~ WL ~ "}")
+  val XNameStart = P("_" | BaseChar | Ideographic)
+  val Name = P(XNameStart ~ NameChar.rep)
+
+  val Char = P(AnyChar)
+  val Char1 = P(!("<" | "&") ~ Char)
+  val CharQ = P(!"\"" ~ Char1)
+  val CharA = P(!"'" ~ Char1)
+  val CharB = P(!"{" ~ Char1)
+
+  val Comment = P("<!--" ~ ((!"-" ~ Char) | ("-" ~ (!"-" ~ Char))).rep ~ "-->")
+
+  val EntityRef = P("&" ~ Name ~ ";")
+  val CharRef = P(
+    "&#" ~ CharIn('0' to '9').rep(1) ~ ";" | "&#x" ~ HexNum ~ ";")
+  val Reference = P(EntityRef | CharRef)
+  val AttValue = P(
+    "\"" ~ (CharQ | Reference).rep ~ "\"" |
+      "'" ~ (CharA | Reference).rep ~ "'" |
+      ScalaExpr
+  )
+  val Attribute = P(Name ~ Eq ~ AttValue)
+  val TagHeader = P("<" ~ Name ~ (WL ~ Attribute).rep ~ WL.?)
+  val CharData = P((!("{" | "]]>" | CharRef) ~ Char1 | "{{").rep(1))
+  val EmptyElemTagEnd = P("/>")
+  val STagEnd = P(">")
+
+  val CDStart = P("<![CDATA[")
+  val CData = P((!"]]>" ~ Char).rep)
+  val CDEnd = P("]]>")
+  val CDSect = P(CDStart ~ CData ~ CDEnd)
+
+  val PITarget = P(!(("X" | "x") ~ ("M" | "m") ~ ("L" | "l")) ~ Name)
+  val PI = P("<?" ~ PITarget ~ (WL ~ (!"?>" ~ Char).rep).? ~ "?>")
+
+  val ETagP = P("</" ~ Name ~ WL.? ~ ">")
+
+  val XmlContent: P0 = P(Element | CDSect | PI | Comment)
+  val Content1 = P(XmlContent | Reference | ScalaExpr)
+  val Content = P((CharData | Content1).rep)
+  val Element = P(TagHeader ~ (EmptyElemTagEnd | STagEnd ~ Content ~ ETag))
+
+  val ETag = P("</" ~ Name ~ WL.? ~ ">")
+
+  val TagPHeader = P("<" ~ Name ~ WL.?)
+  val EmptyElemTagPEnd = P("/>")
+  val STagPEnd = P(">")
+  val ScalaPatterns = P("{" ~ Patterns ~ WL ~ "}")
+  val ElemPattern: P0 = P(
+    TagPHeader ~ (EmptyElemTagPEnd | STagPEnd ~ ContentP ~ ETagP))
+  val ContentP = P(
+    CharData.? ~ ((ElemPattern | ScalaPatterns) ~ CharData.?).rep)
+  val ContentP1 = P(
+    ElemPattern | Reference | CDSect | PI | Comment | ScalaPatterns)
+  val XmlExpr = P(WL ~ XmlContent ~ (WL ~ Element).rep)
+  val XmlPattern = P(WL ~ ElemPattern)
+}
+
+case class RangePosition(from: Int, to: Int)
+
+/** Collects start and end positions of scala expressions inside xml literals.
+  *
+  * Doesn't really parse scala expressions, only reads until the curly brace
+  * balance hits 0.
+  */
+class ScalaExprPositionParser extends fastparse.all.Parser[Unit] {
+  private val splicePositions = Set.newBuilder[RangePosition]
+  def getSplicePositions = splicePositions.result()
+
+  def parseRec(cfg: fastparse.core.ParseCtx, index: Int) = {
+    var current = 0
+    var curlyBraceCount = 1
+    while (curlyBraceCount > 0) {
+      current += 1
+      (cfg.input(index + current): @switch) match {
+        case '}' => curlyBraceCount -= 1
+        case '{' => curlyBraceCount += 1
+        case _ =>
+      }
+    }
```

Comment:
We can't use the parser because the tokenizer doesn't depend on the parser. In cases when the input doesn't parse it can be handy to at least have a functioning tokenizer.

I've fixed the implementation to use the LegacyScanner so that expressions like these

```
<foo>{ "{" + `{`}</foo>
```

are accounted for.